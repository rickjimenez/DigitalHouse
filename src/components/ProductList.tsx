import React, {FC} from 'react';
import {
  Text,
  View,
  StyleSheet,
  FlatList,
  ListRenderItem,
  TouchableOpacity,
  Image,
} from 'react-native';
import {GetProductsResponse} from '../types/api';
import format from 'date-fns/format';
import es from 'date-fns/locale/es';
import {useNavigation} from '@react-navigation/native';
import {HomeProps} from '../types/navigation';

interface Props {
  data: GetProductsResponse[];
  refetch?: () => void;
  isRefetching?: boolean;
}

/**
 * ProductList Component
 *
 * @returns {React.ReactElement} ProductList.
 */
const ProductList: FC<Props> = ({
  data = [],
  refetch = () => {},
  isRefetching = false,
}): JSX.Element => {
  const nav = useNavigation<HomeProps['navigation']>();
  const styles = useStyles();

  const renderItem: ListRenderItem<GetProductsResponse> = ({item, index}) => {
    const lastItem = data.length === index + 1 && styles.lastItem;
    const fistItem = index === 0 && styles.firstItem;
    return (
      <TouchableOpacity
        style={[styles.listItem, lastItem, fistItem]}
        onPress={() => nav.navigate('ProductDetails', {productId: item.id})}>
        <View style={styles.productInfoContainer}>
          <Image style={[styles.thumb]} source={{uri: item.image}} />
          <View>
            <Text style={styles.productName}>{item.product}</Text>
            <Text style={styles.date}>
              {format(new Date(item.createdAt), "dd 'de' MMMM yyyy", {
                locale: es,
              })}
            </Text>
          </View>
        </View>
        <View>
          <Text style={styles.points}>
            <Text
              style={
                item.is_redemption ? styles.negative : styles.positive
              }>{`${item.is_redemption ? '- ' : '+ '}`}</Text>
            {item.points}
          </Text>
        </View>
      </TouchableOpacity>
    );
  };

  return (
    <View style={styles.container}>
      <FlatList
        data={data}
        renderItem={renderItem}
        onRefresh={refetch}
        refreshing={isRefetching}
        showsVerticalScrollIndicator={false}
      />
    </View>
  );
};

const useStyles = () =>
  StyleSheet.create({
    container: {
      backgroundColor: 'white',
      borderRadius: 20,
      flex: 1,
    },
    listItem: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      marginBottom: 10,
      marginHorizontal: 20,
    },
    firstItem: {
      marginTop: 20,
    },
    lastItem: {
      marginBottom: 20,
    },
    productInfoContainer: {
      flexDirection: 'row',
      alignItems: 'center',
    },
    thumb: {
      width: 45,
      height: 45,
      borderRadius: 10,
      marginRight: 10,
    },
    productName: {
      fontSize: 14,
      fontWeight: '800',
    },
    date: {
      fontSize: 12,
      fontWeight: '400',
    },
    points: {
      fontWeight: '800',
      fontSize: 14,
    },
    negative: {
      color: '#FF0000',
    },
    positive: {
      color: '#00B833',
    },
  });

export default ProductList;
