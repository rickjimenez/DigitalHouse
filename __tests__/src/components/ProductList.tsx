import 'react-native';
import {render, screen, fireEvent} from '@testing-library/react-native';
import React from 'react';
import ProductList from '../../../src/components/ProductList';

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('test ProductList Comp', () => {
  test('renders correctly', () => {
    const refetch = jest.fn();
    render(
      <ProductList
        refetch={refetch}
        data={[
          {
            id: '1',
            product: 'product',
            is_redemption: false,
            createdAt: '2022-12-09T06:34:25.607Z',
            image: 'image',
            points: 100,
          },
        ]}
      />,
    );

    const comp = screen.toJSON();
    expect(comp).toMatchSnapshot();
  });

  test('test list', () => {
    const refetch = jest.fn();
    render(
      <ProductList
        refetch={refetch}
        data={[
          {
            id: '1',
            product: 'product',
            is_redemption: true,
            createdAt: '2022-12-09T06:34:25.607Z',
            image: 'image',
            points: 100,
          },
        ]}
      />,
    );
    const itemBtn = screen.getByTestId('item-btn');
    fireEvent.press(itemBtn);
    expect(mockedNavigate).toHaveBeenCalledTimes(1);
  });
});
