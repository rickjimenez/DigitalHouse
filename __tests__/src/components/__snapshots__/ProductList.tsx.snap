// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Button Comp renders correctly 1`] = `
[
  <Text
    style={
      {
        "color": "#9B9898",
        "fontSize": 14,
        "fontWeight": "800",
        "marginBottom": 10,
      }
    }
  >
    TUS PUNTOS
  </Text>,
  <View
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 20,
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      testID="list"
      viewabilityConfigCallbackPairs={[]}
    >
      <RCTRefreshControl />
      <View />
    </RCTScrollView>
  </View>,
]
`;

exports[`test ProductList Comp renders correctly 1`] = `
[
  <Text
    style={
      {
        "color": "#9B9898",
        "fontSize": 14,
        "fontWeight": "800",
        "marginBottom": 10,
      }
    }
  >
    TUS PUNTOS
  </Text>,
  <View
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 20,
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      data={
        [
          {
            "createdAt": "2022-12-09T06:34:25.607Z",
            "id": "1",
            "image": "image",
            "is_redemption": false,
            "points": 100,
            "product": "product",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[MockFunction]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[MockFunction]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={[]}
      testID="list"
      viewabilityConfigCallbackPairs={[]}
    >
      <RCTRefreshControl />
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 20,
                "marginHorizontal": 20,
                "marginTop": 20,
                "opacity": 1,
              }
            }
            testID="item-btn"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Image
                source={
                  {
                    "cache": "default",
                    "uri": "image",
                  }
                }
                style={
                  [
                    {
                      "borderRadius": 10,
                      "height": 45,
                      "marginRight": 10,
                      "width": 45,
                    },
                  ]
                }
              />
              <View>
                <Text
                  style={
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                    }
                  }
                >
                  product
                </Text>
                <Text
                  style={
                    {
                      "fontSize": 12,
                      "fontWeight": "400",
                    }
                  }
                >
                  09 de diciembre 2022
                </Text>
              </View>
            </View>
            <View>
              <Text
                style={
                  {
                    "fontSize": 14,
                    "fontWeight": "800",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#00B833",
                    }
                  }
                >
                  + 
                </Text>
                100
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>,
]
`;
